test/case/ietf_system/upgrade/test.py

```diff
@@ -53,17 +96,49 @@
             })
 
         with test.step("Wait for upgrade to finish"):
-            for _ in range(600):
+            for left in range(600):
                 oper = target.get_dict("/system-state/software")
                 installer = oper["system-state"]["software"]["installer"]
                 if installer["operation"] == "idle":
                     if "last-error" in installer:
                         print("Install failed:", installer["last-error"])
                         test.fail()
 
-                    test.succeed()
-
+                    break
+                time.sleep(1)
+                if left >= 600:
+                    print("Timeout, last state:", oper)
+                    test.fail()
                 time.sleep(1)
 
-            print("Timeout, last state:", oper)
-            test.fail()
+            with test.step("Verify boot order has changed and reboot"):
+                assert(old_bootorder != bootloader.get_boot_order())
+                target.reboot()
+
+                if not wait_boot(target, env):
+                    test.fail()
+                target = env.attach("target", "mgmt", "netconf")
+
+
+        with test.step("Verify that the partition is the booted"):
+            should_boot=bootloader.get_boot_order().split()[0]
+            oper = target.get_dict("/system-state/software")
+            booted = oper["system-state"]["software"]["booted"]
+            print(f"Should boot: {should_boot}, booted: {booted}")
+            assert(booted == should_boot)
+
+        with test.step("Restore boot order to original configured"):
+            print(f"Restore boot order to {old_bootorder}")
+            if bootloader.set_boot_order(old_bootorder) != 0:
+                test.fail()
+            target = env.attach("target", "mgmt", "netconf")
+            target.reboot()
+            if not wait_boot(target, env):
+                test.fail()
+            target = env.attach("target", "mgmt", "netconf")
+
+        with test.step("Verify the boot order is the orignal configured"):
+            # Wait for sshd to start.
+            until(lambda: old_bootorder == bootloader.get_boot_order(), attempts=200)
```

Comment:
```suggestion
            order = bootloader.get_boot_order()
            assert order == old_bootorder, f"Unexpected bootorder: {repr(order)}" 
```
`until` seems counter intuitive here, since the boot order is not a volatile thing. Do we need more assurances from the SSH connection? Should we add a long `-oConnectTimeout 300` or something?

Also, if this ever happens, I think we want to know which state we're leaving the system in